rust/src/shards/gui/layouts/table.rs

```diff
@@ -0,0 +1,341 @@
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      rows: ParamVar::default(),
+      columns: ParamVar::default(),
+      striped: ParamVar::default(),
+      resizable: ParamVar::default(),
+      row_index: ParamVar::default(),
+      shards: Vec::new(),
+      exposing: Vec::new(),
+      should_expose: false,
+    }
+  }
+}
+
+impl Shard for Table {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.Table")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.Table-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.Table"
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("Table layout."))
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANYS_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANYS_TYPES
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&TABLE_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => {
+        let seq = Seq::try_from(value)?;
+        for shard in seq.iter() {
+          let mut s = ShardsVar::default();
+          s.set_param(&shard).unwrap();
+          self.shards.push(s);
+        }
+        Ok(self.rows.set_param(value))
+      }
+      1 => Ok(self.columns.set_param(value)),
+      2 => Ok(self.striped.set_param(value)),
+      3 => Ok(self.resizable.set_param(value)),
+      4 => Ok(self.row_index.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.rows.get_param(),
+      1 => self.columns.get_param(),
+      2 => self.striped.get_param(),
+      3 => self.resizable.get_param(),
+      4 => self.row_index.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    util::require_parents(&mut self.requiring, &self.parents);
+
+    Some(&self.requiring)
+  }
+
+  fn exposedVariables(&mut self) -> Option<&ExposedTypes> {
+    self.exposing.clear();
+
+    let mut exposed = false;
+    for s in &self.shards {
+      exposed |= util::expose_contents_variables(&mut self.exposing, s);
+    }
+
+    if self.row_index.is_variable() {
+      let exp_info = ExposedInfo {
+        exposedType: common_type::int,
+        name: self.row_index.get_name(),
+        help: cstr!("The exposed row index variable").into(),
+        ..ExposedInfo::default()
+      };
+
+      exposed = true;
+      self.exposing.push(exp_info);
+    }
+
+    if exposed {
+      Some(&self.exposing)
+    } else {
+      None
+    }
+  }
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    for s in &mut self.shards {
+      s.compose(data)?;
```

Comment:
I think you have to modify the input type here to match the sequence elements

Reply:
I will try.

Reply:
I tried to modify the `inputType` before composing but I'm hit with an error:
> [trace] [2022-10-20 14:14:37.665] [T-42112] [runtime.cpp::2955] composeShards failed: basic_string: construction from null is not valid
[info] [2022-10-20 14:14:37.665] [T-42112] [runtime.cpp::3007] Compose failed with error basic_string: construction from null is not valid
[error] [2022-10-20 14:14:37.665] [T-42112] [runtime.cpp::1031] Error composing shard: Composition failed., wire: main-wire
[trace] [2022-10-20 14:14:37.665] [T-42112] [runtime.cpp::2955] composeShards failed: Composition failed.

```rust
fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
  let mut data = *data;
  //...
  let input_type = data.inputType;
  let slice = unsafe {
    let ptr = input_type.details.seqTypes.elements;
    std::slice::from_raw_parts(ptr, input_type.details.seqTypes.len as usize)
  };
  let element_type = if !slice.is_empty() {
    slice[0]
  } else {
    common_type::any
  };

  for s in &mut self.shards {
    data.inputType = element_type;
    s.compose(&data)?;
  }

  // Always passthrough the input
  Ok(input_type)
}
```

Not sure why I get an error about a string, while I'm only manipulating types.